src/util/bitset.h

```diff
@@ -0,0 +1,463 @@
+        return *this;
+    }
+    /** Construct a bitset with the singleton i. */
+    static constexpr IntBitSet Singleton(unsigned i) noexcept {
+        return IntBitSet(I(1U) << i);
+    }
+    /** Construct a bitset with bits 0..count-1 (inclusive) set to 1. */
+    static constexpr IntBitSet Fill(unsigned count) noexcept {
+        IntBitSet ret;
+        if (count) ret.m_val = I(~I{0}) >> (MAX_SIZE - count);
+        return ret;
+    }
+    /** Compute the number of 1 bits in the bitset. */
+    constexpr unsigned Count() const noexcept { return PopCount(m_val); }
+    /** Return the number of bits that this object holds. */
+    static constexpr unsigned Size() noexcept { return MAX_SIZE; }
+    /** Set a bit to 1. */
+    constexpr void Set(unsigned pos) noexcept { m_val |= I{1U} << pos; }
+    /** Set a bit to the specified value. */
+    constexpr void Set(unsigned pos, bool val) noexcept { m_val = (m_val & ~I(I{1U} << pos)) | (I(val) << pos); }
+    /** Set a bit to 0. */
+    constexpr void Reset(unsigned pos) noexcept { m_val &= ~I(I{1U} << pos); }
+    /** Retrieve a bit at the given position. */
+    constexpr bool operator[](unsigned pos) const noexcept { return (m_val >> pos) & 1U; }
+    /** Check if all bits are 0. */
+    constexpr bool None() const noexcept { return m_val == 0; }
+    /** Check if any bits are 1. */
+    constexpr bool Any() const noexcept { return m_val != 0; }
+    /** Return an object that iterates over all 1 bits (++ and * only allowed when != end()). */
+    constexpr Iterator begin() const noexcept { return Iterator(m_val); }
+    /** Return a dummy object to compare Iterators with. */
+    constexpr IteratorEnd end() const noexcept { return IteratorEnd(); }
+    /** Find the first element (requires Any()). */
+    constexpr unsigned First() const noexcept { return std::countr_zero(m_val); }
+    /** Find the last element (requires Any()). */
+    constexpr unsigned Last() const noexcept { return std::bit_width(m_val) - 1; }
+    /** Set this object's bits to be the binary AND between respective bits from this and a. */
+    constexpr IntBitSet& operator|=(const IntBitSet& a) noexcept { m_val |= a.m_val; return *this; }
+    /** Set this object's bits to be the binary OR between respective bits from this and a. */
+    constexpr IntBitSet& operator&=(const IntBitSet& a) noexcept { m_val &= a.m_val; return *this; }
+    /** Set this object's bits to be the binary AND NOT between respective bits from this and a. */
+    constexpr IntBitSet& operator/=(const IntBitSet& a) noexcept { m_val &= ~a.m_val; return *this; }
+    /** Set this object's bits to be the binary XOR between respective bits from this as a. */
+    constexpr IntBitSet& operator^=(const IntBitSet& a) noexcept { m_val ^= a.m_val; return *this; }
+    /** Check if the intersection between two sets is non-empty. */
+    friend constexpr bool operator&&(const IntBitSet& a, const IntBitSet& b) noexcept { return a.m_val & b.m_val; }
```

Comment:
I know you like dense code, but...

I suspect function names would make reviewing uses of this easier than operators. Especially for the non-obvious ones like `&&` that `std::bitset` doesn't support.

I know that personally I'd have an easier time reviewing `a.intersects(b)`.